op-challenger/challenger/traversal.go

```diff
@@ -0,0 +1,247 @@
+package challenger
+
+import (
+	"context"
+	"errors"
+	"math/big"
+	"sync"
+	"time"
+
+	"github.com/ethereum/go-ethereum"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/log"
+
+	"github.com/ethereum-optimism/optimism/op-node/client"
+	"github.com/ethereum-optimism/optimism/op-node/eth"
+	"github.com/ethereum-optimism/optimism/op-node/sources"
+	"github.com/ethereum-optimism/optimism/op-node/sources/caching"
+	"github.com/ethereum-optimism/optimism/op-service/backoff"
+)
+
+var (
+	// ErrAlreadyStarted is returned when the log traversal has already been started.
+	ErrAlreadyStarted = errors.New("already started")
+	// ExponentialBackoff is the default backoff strategy.
+	ExponentialBackoff = backoff.Exponential()
+)
+
+// logTraversal implements LogTraversal.
+type logTraversal struct {
+	log             log.Logger
+	client          MinimalEthClient
+	query           *ethereum.FilterQuery
+	quit            chan struct{}
+	mutex           sync.Mutex
+	lastBlockNumber *big.Int
+	started         bool
+}
+
+//go:generate mockery --name MinimalEthClient --output ./mocks/
+type MinimalEthClient interface {
+	InfoByHash(ctx context.Context, hash common.Hash) (eth.BlockInfo, error)
+	InfoByNumber(ctx context.Context, number uint64) (eth.BlockInfo, error)
+	FetchReceipts(ctx context.Context, blockHash common.Hash) (eth.BlockInfo, types.Receipts, error)
+	SubscribeNewHead(ctx context.Context, ch chan<- *types.Header) (ethereum.Subscription, error)
+}
+
+// NewMinimalEthClient creates a [sources.EthClient] with a minimal interface.
+func NewMinimalEthClient(client client.RPC, log log.Logger, metrics caching.Metrics, config *sources.EthClientConfig) (MinimalEthClient, error) {
```

Comment:
no need for this - as `*sources.EthClient` satisfies `MinimalEthClient` you can directly replace `NewMinimalEthClient ` with `sources.NewEthClient`. You're following the language idiom of accepting interfaces, but not the idiom of returning structs. In this case there is not really any benefit of returning a interface here.